### PR TITLE
Inital commit for the auto generation of the api docs

### PR DESCRIPTION
**Pull request by jrbenny35**
Just a rough draft really. 

@davehunt r?

Also after this I will do the User Guide.

### CODE REVIEW

File: docs/api.rst [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  :inherited-members:
+
+
+Browser Model
+=============
+
+.. py:module:: foxpuppet.windows.browser
+
+.. autoclass:: BrowserWindow
+  :inherited-members:
```
Comment (davehunt) [inline docs/api.rst]: The structure should be a heading for windows and then a section for each window implementation. We shouldn't need to document the abstract base class.

```
# Windows
## Window Manager
## Browser Window
```

File: docs/api.rst [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+Base Window Model
+=================
+
+.. py:module:: foxpuppet.windows.base
+
+.. autoclass:: BaseWindow
+  :inherited-members:
+
+
+Browser Model
```
Comment (davehunt) [inline docs/api.rst]: Remove 'Model' from the headings.

File: foxpuppet/foxpuppet.py [hunk 3]
```diff
@@ -8,7 +8,11 @@
 
 
 class FoxPuppet(object):
-    """ Class that sets up the api for interacting with the Firefox browser.
+    """
+        Class that sets up the interface for interacting with the Firefox
+        browser.
+
+        :param selenium: A selenium object
```
Comment (davehunt) [inline foxpuppet/foxpuppet.py]: We should use intersphinx and link the param type to the official Selenium documentation as we do with PyPOM: https://github.com/mozilla/PyPOM/blob/master/docs/conf.py#L41

File: foxpuppet/windows/browser.py [hunk 4]
```diff
@@ -37,6 +41,8 @@ def open_window(self, private=False):
 
         :param private: Optional parameter to open a private browsing window.
                         Defaults to False.
+
+        :returns: A BrowserWindow object of the opened window.
```
Comment (davehunt) [inline foxpuppet/windows/browser.py]: We should include the return types, linking to the relevant documentation.

File: foxpuppet/windows/manager.py [hunk 5]
```diff
@@ -5,14 +5,23 @@
 
 class WindowManager(object):
 
+    """
+        A window manager that controls the creation of window objects
+        for interaction.
```
Comment (davehunt) [inline foxpuppet/windows/manager.py]: I don't think this description is particularly useful. Perhaps say that it provides access to all open windows. The user doesn't need to know that it creates the window objects.

File: foxpuppet/windows/manager.py [hunk 6]
```diff
@@ -5,14 +5,23 @@
...
         self.selenium = selenium
 
     @property
     def windows(self):
         """
-        Sets all current window handles to appropriate window instances
-        :returns: A list of BrowserWindow Instances
+        Sets all current windows to appropriate window instances.
+
+        :returns: A list of Window Instances
```
Comment (davehunt) [inline foxpuppet/windows/manager.py]: Include the return types.

**Comment by jrbenny35** [comment]
@davehunt Thanks for the review. When you say push to a branch on the mozilla fork, you mean create a branch on the Mozilla fork of foxpuppet?

File: docs/api.rst (hunk 1 above)
Comment (jrbenny35) [inline docs/api.rst]: Would we have to say anything about the 'Windows' directory?

File: foxpuppet/windows/browser.py (hunk 4 above)
Comment (jrbenny35) [inline foxpuppet/windows/browser.py]: Linking to what specifically?

**Comment by jrbenny35** [comment]
I pushed to a branch here but the docs aren't rendering the new changes. I can do a build locally and get them however.

**Comment by davehunt** [comment]
I've enabled the sphix_auto_docs branch as a protected version here: http://foxpuppet.readthedocs.io/en/sphinx_auto_docs/api.html. As you can see the autogenerated docs are not showing up yet. Please close this pull, push only to the branch on the mozilla fork for now, and when you're ready open a new pull request from the new branch.

File: docs/api.rst (hunk 1 above)
Comment (davehunt) [inline docs/api.rst]: I don't think so, users of a Python client will not think of this as a directory, but as a package/module.

File: foxpuppet/foxpuppet.py (hunk 3 above)
Comment (jrbenny35) [inline foxpuppet/foxpuppet.py]: Is there a good example of this within a docstring? It would work within an rst file but not within a py file.

**Comment by jrbenny35** [comment]
Closing.

**Pull request closed without merging**